tfhe-zk-pok/src/curve_api/bls12_381.rs

```diff
@@ -36,6 +36,38 @@ fn bigint_to_bytes(x: [u64; 6]) -> [u8; 6 * 8] {
 mod g1 {
     use super::*;
 
+    #[derive(
+        Copy,
+        Clone,
+        Debug,
+        PartialEq,
+        Eq,
+        Serialize,
+        Deserialize,
+        Hash,
+        CanonicalSerialize,
+        CanonicalDeserialize,
+    )]
+    #[repr(transparent)]
+    pub struct G1Affine {
+        #[serde(serialize_with = "ark_se", deserialize_with = "ark_de")]
+        pub(crate) inner: ark_bls12_381::g1::G1Affine,
+    }
+
+    impl G1Affine {
+        pub fn multi_mul_scalar(bases: &[Self], scalars: &[Zp]) -> G1 {
+            G1 {
+                inner: ark_bls12_381::g1::G1Projective::msm(
+                    unsafe {
+                        &*(bases as *const [G1Affine] as *const [ark_bls12_381::g1::G1Affine])
+                    },
+                    unsafe { &*(scalars as *const [Zp] as *const [ark_bls12_381::Fr]) },
+                )
+                .unwrap(),
```

Comment:
why the need for unsafe ? and what are the safety requirements ?

Reply:
this is essentially interpreting the wrapper struct `Zp(ark_bls12_381::Fr)` as its contents `Fr`. this is safe since it's `repr(transparent)`. we can remove the unsafety but that would require an allocation + copy which has non negligible cost, at least for the bases

Reply:
I'm fine with the unsafe as long as the comment about the repr transparent is given in the code :)

Reply:
also if the wrapper proves to be an issue it can be removed down the line, the BLS446 is the final choice now

Reply:
added a comment `// SAFETY: interpreting a `repr(transparent)` pointer as its contents.`

Reply:
thanks !